beacon-chain/sync/initial-sync/service.go

```diff
@@ -293,332 +258,65 @@ func (s *InitialSync) checkInMemoryBlocks() {
-	return false
-}
-
-func (s *InitialSync) processBlockAnnounce(msg p2p.Message) {
-	ctx, span := trace.StartSpan(msg.Ctx, "beacon-chain.sync.initial-sync.processBlockAnnounce")
-	defer span.End()
-	data := msg.Data.(*pb.BeaconBlockAnnounce)
-	recBlockAnnounce.Inc()
-
-	if s.reqState {
-		if err := s.requestStateFromPeer(ctx, s.stateRootOfHighestObservedSlot[:], msg.Peer); err != nil {
-			log.Errorf("Could not request state from peer %v", err)
-		}
-		return
-	}
-
-	if data.SlotNumber > s.highestObservedSlot {
-		s.highestObservedSlot = data.SlotNumber
-	}
-
-	s.requestBatchedBlocks(s.currentSlot+1, s.highestObservedSlot)
-	log.Debugf("Successfully requested the next block with slot: %d", data.SlotNumber-params.BeaconConfig().GenesisSlot)
-}
-
-// processBlock is the main method that validates each block which is received
-// for initial sync. It checks if the blocks are valid and then will continue to
-// process and save it into the db.
-func (s *InitialSync) processBlock(ctx context.Context, block *pb.BeaconBlock, peerID peer.ID) {
-	ctx, span := trace.StartSpan(ctx, "beacon-chain.sync.initial-sync.processBlock")
-	defer span.End()
-	recBlock.Inc()
-	if block.Slot > s.highestObservedSlot {
-		s.highestObservedSlot = block.Slot
-	}
-
-	if block.Slot < s.currentSlot {
-		return
-	}
-
-	// requesting beacon state if there is no saved state.
-	if s.reqState {
-		if err := s.requestStateFromPeer(s.ctx, s.stateRootOfHighestObservedSlot[:], peerID); err != nil {
-			log.Errorf("Could not request beacon state from peer: %v", err)
-		}
-		return
-	}
-	// if it isn't the block in the next slot we check if it is a skipped slot.
-	// if it isn't skipped we save it in memory.
-	if block.Slot != (s.currentSlot + 1) {
-		// if parent exists we validate the block.
-		if s.doesParentExist(block) {
-			if err := s.validateAndSaveNextBlock(ctx, block); err != nil {
-				// Debug error so as not to have noisy error logs
-				if strings.HasPrefix(err.Error(), debugError) {
-					log.Debug(strings.TrimPrefix(err.Error(), debugError))
-					return
-				}
-				log.Errorf("Unable to save block: %v", err)
-			}
-			return
-		}
-		s.mutex.Lock()
-		defer s.mutex.Unlock()
-		if _, ok := s.inMemoryBlocks[block.Slot]; !ok {
-			s.inMemoryBlocks[block.Slot] = block
-		}
-		return
-	}
-
-	if err := s.validateAndSaveNextBlock(ctx, block); err != nil {
-		// Debug error so as not to have noisy error logs
-		if strings.HasPrefix(err.Error(), debugError) {
-			log.Debug(strings.TrimPrefix(err.Error(), debugError))
+// listenForNewBlocks listens for block announcements beyond the canonical head slot that may
+// be received during initial sync - we must process these blocks to catch up with peers.
+func (s *InitialSync) listenForNewBlocks() {
+	blockAnnounceSub := s.p2p.Subscribe(&pb.BeaconBlockAnnounce{}, s.blockAnnounceBuf)
+	defer func() {
+		blockAnnounceSub.Unsubscribe()
+		close(s.blockAnnounceBuf)
+	}()
+	for {
+		select {
+		case <-s.ctx.Done():
```

Comment:
if we start normal sync, does this still listen for blocks or is this routine still running ?